### PR TITLE
Remove bundles when deduplicating entry assets that are referenced by their descendants

### LINKED ISSUE #5181: @parcel/bundler-default: Got unexpected undefined
# 🐛 bug report

While testing out parcel v2 I ran into a fatal build error with an internal-looking failure.

```bash

ℹ️ Server running at http://localhost:1234
🚨 Build failed.
@parcel/bundler-default: Got unexpected undefined
Error: Got unexpected undefined
    at nullthrows (/node_modules/nullthrows/nullthrows.js:7:15)
    at BundleGraph.getBundleGroupsContainingBundle (/node_modules/@parcel/core/lib/BundleGraph.js:718:70)
    at /node_modules/@parcel/core/lib/BundleGraph.js:393:64
    at Array.every (<anonymous>)
    at BundleGraph.removeBundleGroup (/node_modules/@parcel/core/lib/BundleGraph.js:393:43)
    at MutableBundleGraph.removeBundleGroup (/node_modules/@parcel/core/lib/public/MutableBundleGraph.js:254:41)
    at Object.optimize (/node_modules/@parcel/bundler-default/lib/DefaultBundler.js:426:21)
    at BundlerRunner.bundle (/node_modules/@parcel/core/lib/BundlerRunner.js:276:21)
    at async Parcel.build (/node_modules/@parcel/core/lib/Parcel.js:579:25)
    at async Parcel.startNextBuild (/node_modules/@parcel/core/lib/Parcel.js:481:21)

```

## 🎛 Configuration (.babelrc, package.json, cli command)

```json
{
	"name": "dev",
	"version": "0.0.1",
	"license": "MIT",
	"scripts": {
		"serve": "parcel './src/pages/**.*'",
		"build": "parcel build './src/pages/**.html'"
	},
	"devDependencies": {
		"parcel": "^2.0.0-beta.1"
	},
	"dependencies": {
		"jquery": "^3.5.1"
	}
}
```

## 🤔 Expected Behavior

Successful execution.

## 😯 Current Behavior

A fatal build error with an internal-looking failure.

## 🌍 Your Environment

| Software         | Version(s) |
| ---------------- | ---------- |
| Parcel           | 2.0.0-beta.1
| Node             | v14.12.0
| Yarn         | 1.22.5
| Operating System | macOS 10.13.6

### PR DESCRIPTION
**Pull request by wbinnssmith**
Closes #5181

This fixes an issue where a bundle would not be properly removed if its entry asset was referenced by a descendant asset that was deduplicated. Because we only checked if the root of the asset graph to be removed was the entry [0], the untyped edge would not be removed even though the contains edge was.

This also updates an assertion that runs after removing a bundle group to only happen if `removeBundleGroup` was not re-entered: it's possible for `removeBundleGroup` to remove bundles if they only belonged to that group, and for `removeBundle` to call `removeBundleGroup` if it leaves the bundle group entry. This is covered by the same test case.

This also fixes an oversight in `removeAssetGraph` where a reference edge was _added_ between a bundle and a dependency if it already existed, rather than removing it.

Test Plan: Added integration test

[0] https://github.com/parcel-bundler/parcel/blob/806cc65e0e430b1120cdcf1676e2f58738326d30/packages/core/core/src/BundleGraph.js#L408-L410

### CODE REVIEW

**Comment by wbinnssmith** [comment]
cc @marcins

File: packages/core/integration-tests/test/html.js [hunk 1]
```diff
@@ -1539,16 +1539,16 @@ describe('html', function() {
     });
 
     let html = await outputFS.readFile(path.join(distDir, 'a.html'), 'utf8');
-    assert.equal(html.match(/<script/g).length, 3);
+    assert.equal(html.match(/<script/g).length, 2);
```
Comment (wbinnssmith) [inline packages/core/integration-tests/test/html.js:1542]: The contents of the additional bundle before was empty (aside from the sourcemap comment). Looks like it was a case of this issue: a shared bundle was never properly removed from the graph.

**Comment by parcel-benchmark** [comment]
## Benchmark Results
<details><summary>Kitchen Sink ✅</summary><p>

#### Timings

| Description | Time | Difference |
| --- | --- | --- |
| Cold | 8.97s | -45.00ms |
| Cached | 560.00ms | +16.00ms |

#### Cold Bundles

| Bundle | Size | Difference | Time | Difference |
| --- | --- | --- | --- | --- |
| dist/legacy/parcel.d5807e82.webp | 102.94kb | +0.00b | 626.00ms | +89.00ms ⚠️ |
| dist/modern/parcel.d5807e82.webp | 102.94kb | +0.00b | 537.00ms | -57.00ms 🚀 |
| dist/modern/index.b43fcc94.js | 1.12kb | +0.00b | 1.00s | -345.00ms 🚀 |
| dist/modern/index.html | 701.00b | +0.00b | 627.00ms | -479.00ms 🚀 |
| dist/modern/index.0bc656bc.css | 77.00b | +0.00b | 1.00s | -350.00ms 🚀 |

#### Cached Bundles

| Bundle | Size | Difference | Time | Difference |
| --- | --- | --- | --- | --- |
| dist/legacy/parcel.d5807e82.webp | 102.94kb | +0.00b | 89.00ms | +31.00ms ⚠️ |
| dist/modern/parcel.d5807e82.webp | 102.94kb | +0.00b | 66.00ms | -13.00ms 🚀 |
| dist/legacy/index.76ce4b0c.js | 1.12kb | +0.00b | 89.00ms | +9.00ms ⚠️ |
| dist/modern/index.b43fcc94.js | 1.12kb | +0.00b | 88.00ms | +8.00ms ⚠️ |
| dist/legacy/index.html | 701.00b | +0.00b | 89.00ms | +6.00ms ⚠️ |
| dist/modern/index.html | 701.00b | +0.00b | 89.00ms | +9.00ms ⚠️ |
| dist/legacy/index.44edca46.css | 77.00b | +0.00b | 89.00ms | +9.00ms ⚠️ |
| dist/modern/index.0bc656bc.css | 77.00b | +0.00b | 89.00ms | +9.00ms ⚠️ |

</p></details>

<details><summary>React HackerNews ✅</summary><p>

#### Timings

| Description | Time | Difference |
| --- | --- | --- |
| Cold | 33.24s | +1.18s |
| Cached | 1.23s | +85.00ms ⚠️ |

#### Cold Bundles

| Bundle | Size | Difference | Time | Difference |
| --- | --- | --- | --- | --- |
| dist/PermalinkedComment.0a19c7e4.js | 4.24kb | +0.00b | 1.29s | +323.00ms ⚠️ |
| dist/UserProfile.b2960402.js | 1.78kb | +0.00b | 1.29s | +323.00ms ⚠️ |
| dist/NotFound.f3269ae7.js | 542.00b | +0.00b | 370.00ms | +104.00ms ⚠️ |
| dist/logo.24c8bf9e.png | 274.00b | +0.00b | 649.00ms | +517.00ms ⚠️ |

#### Cached Bundles

*No bundle changes detected.*

</p></details>

<details><summary>AtlasKit Editor ✅</summary><p>

#### Timings

| Description | Time | Difference |
| --- | --- | --- |
| Cold | 2.64m | -2.25s |
| Cached | 3.36s | +205.00ms ⚠️ |

#### Cold Bundles

| Bundle | Size | Difference | Time | Difference |
| --- | --- | --- | --- | --- |
| dist/Toolbar.0c83bb49.js | 100.68kb | +0.00b | 16.37s | -892.00ms 🚀 |
| dist/media-viewer.ebcb5e1e.js | 73.43kb | +0.00b | 11.74s | -1.21s 🚀 |
| dist/esm.e3419f0a.js | 27.83kb | +0.00b | 7.88s | -2.34s 🚀 |
| dist/DatePicker.faca45d0.js | 21.38kb | +0.00b | 9.25s | -926.00ms 🚀 |
| dist/smartMediaEditor.21f91ed8.js | 16.96kb | +0.00b | 14.36s | -2.90s 🚀 |
| dist/dropzone.3ae9a030.js | 16.15kb | +0.00b | 13.12s | -3.67s 🚀 |
| dist/workerHasher.bc6346e1.js | 11.90kb | +0.00b | 8.43s | +887.00ms ⚠️ |
| dist/card.3a4028bd.js | 5.74kb | +0.00b | 11.50s | -1.17s 🚀 |
| dist/media-viewer.f15e56f4.js | 4.01kb | +0.00b | 11.67s | -1.27s 🚀 |
| dist/EmojiPickerComponent.4991ddd4.js | 3.68kb | +0.00b | 9.12s | +1.64s ⚠️ |
| dist/dropzone.d1e8004d.js | 3.46kb | +0.00b | 13.07s | +849.00ms ⚠️ |
| dist/index.ece3a13e.css | 3.46kb | +0.00b | 43.39s | +25.48s ⚠️ |
| dist/ResourcedEmojiComponent.5046c0ff.js | 2.19kb | +0.00b | 10.80s | +3.09s ⚠️ |
| dist/feedback.facaac2e.js | 1.87kb | +0.00b | 9.23s | +4.02s ⚠️ |
| dist/16.89d45d15.js | 1.80kb | +0.00b | 4.62s | +389.00ms ⚠️ |
| dist/workerHasher.31326e89.js | 1.77kb | +0.00b | 8.39s | -1.82s 🚀 |
| dist/list-number.d4568b49.js | 1.69kb | +0.00b | 4.91s | -355.00ms 🚀 |
| dist/link.5be155be.js | 1.54kb | +0.00b | 4.91s | -355.00ms 🚀 |
| dist/heading6.5a6082c8.js | 1.53kb | +0.00b | 9.23s | +4.03s ⚠️ |
| dist/heading3.1c92891c.js | 1.52kb | +0.00b | 6.65s | +1.46s ⚠️ |
| dist/16.07ccad90.js | 1.51kb | +0.00b | 4.46s | -264.00ms 🚀 |
| dist/16.00f25203.js | 1.47kb | +0.00b | 4.62s | +389.00ms ⚠️ |
| dist/16.00100306.js | 1.44kb | +0.00b | 5.19s | +524.00ms ⚠️ |
| dist/16.bdd61d2d.js | 1.41kb | +0.00b | 4.63s | -268.00ms 🚀 |
| dist/heading5.88bc6249.js | 1.41kb | +0.00b | 6.65s | +1.46s ⚠️ |
| dist/expand.4a483eb0.js | 1.39kb | +0.00b | 9.23s | -945.00ms 🚀 |
| dist/16.c9db7e6b.js | 1.32kb | +0.00b | 5.19s | +524.00ms ⚠️ |
| dist/mention.c44b48ed.js | 1.30kb | +0.00b | 5.44s | +501.00ms ⚠️ |
| dist/heading4.a2524c70.js | 1.29kb | +0.00b | 6.65s | +1.46s ⚠️ |
| dist/16.acff901d.js | 1.27kb | +0.00b | 4.88s | +414.00ms ⚠️ |
| dist/16.015dfb9d.js | 1.26kb | +0.00b | 4.45s | +226.00ms ⚠️ |
| dist/16.828b958e.js | 1.26kb | +0.00b | 4.46s | -264.00ms 🚀 |
| dist/16.b9b1643b.js | 1.26kb | +0.00b | 4.46s | -264.00ms 🚀 |
| dist/16.ec84a250.js | 1.26kb | +0.00b | 4.62s | +389.00ms ⚠️ |
| dist/media-card-analytics-error-boundary.ff0177b3.js | 1.25kb | +0.00b | 11.50s | +676.00ms ⚠️ |
| dist/action.01ec8983.js | 1.24kb | +0.00b | 5.19s | +524.00ms ⚠️ |
| dist/list.2dfe3f18.js | 1.19kb | +0.00b | 4.91s | -355.00ms 🚀 |
| dist/panel.fc70749c.js | 1.11kb | +0.00b | 5.07s | -537.00ms 🚀 |
| dist/media-card-analytics-error-boundary.b884d580.js | 1019.00b | +0.00b | 16.68s | +2.70s ⚠️ |
| dist/simpleHasher.79137cb4.js | 767.00b | +0.00b | 8.39s | +908.00ms ⚠️ |

#### Cached Bundles

| Bundle | Size | Difference | Time | Difference |
| --- | --- | --- | --- | --- |
| dist/index.1cee198f.js | 2.46mb | +6.00b ⚠️ | 664.00ms | +83.00ms ⚠️ |
| dist/pdfRenderer.f5a3f4a7.js | 1.11mb | +4.00b ⚠️ | 725.00ms | +60.00ms ⚠️ |
| dist/editorView.ea13a2b7.js | 588.08kb | +0.00b | 733.00ms | +43.00ms ⚠️ |
| dist/popup.119ff481.js | 171.12kb | +0.00b | 726.00ms | +41.00ms ⚠️ |
| dist/EmojiPickerComponent.c4358f76.js | 139.84kb | +0.00b | 709.00ms | +46.00ms ⚠️ |
| dist/Toolbar.0c83bb49.js | 100.68kb | +0.00b | 733.00ms | +43.00ms ⚠️ |
| dist/media-viewer.ebcb5e1e.js | 73.43kb | +0.00b | 716.00ms | +51.00ms ⚠️ |
| dist/card.df70c9b7.js | 54.67kb | +0.00b | 716.00ms | +61.00ms ⚠️ |
| dist/card.1d931b4c.js | 52.43kb | +0.00b | 729.00ms | +44.00ms ⚠️ |
| dist/component.ad27194c.js | 31.05kb | +0.00b | 666.00ms | +86.00ms ⚠️ |
| dist/esm.e3419f0a.js | 27.83kb | +0.00b | 710.00ms | +47.00ms ⚠️ |
| dist/component.f25dcadb.js | 22.80kb | +0.00b | 627.00ms | +32.00ms ⚠️ |
| dist/DatePicker.faca45d0.js | 21.38kb | +0.00b | 709.00ms | +46.00ms ⚠️ |
| dist/smartMediaEditor.21f91ed8.js | 16.96kb | +0.00b | 733.00ms | +43.00ms ⚠️ |
| dist/js.8167e2d5.js | 16.55kb | +0.00b | 627.00ms | +36.00ms ⚠️ |
| dist/dropzone.3ae9a030.js | 16.15kb | +0.00b | 725.00ms | +41.00ms ⚠️ |
| dist/ui.15dd7acd.js | 14.30kb | +0.00b | 718.00ms | +63.00ms ⚠️ |
| dist/workerHasher.bc6346e1.js | 11.90kb | +0.00b | 718.00ms | +43.00ms ⚠️ |
| dist/component.677fad45.js | 6.23kb | +0.00b | 641.00ms | +42.00ms ⚠️ |
| dist/card.3a4028bd.js | 5.74kb | +0.00b | 716.00ms | +51.00ms ⚠️ |
| dist/media-viewer.f15e56f4.js | 4.01kb | +0.00b | 725.00ms | +40.00ms ⚠️ |
| dist/EmojiPickerComponent.4991ddd4.js | 3.68kb | +0.00b | 718.00ms | +63.00ms ⚠️ |
| dist/png-chunks-extract.608f438d.js | 3.56kb | +0.00b | 627.00ms | +32.00ms ⚠️ |
| dist/dropzone.d1e8004d.js | 3.46kb | +0.00b | 725.00ms | +41.00ms ⚠️ |
| dist/index.ece3a13e.css | 3.46kb | +0.00b | 736.00ms | +41.00ms ⚠️ |
| dist/Modal.08282ab6.js | 3.23kb | +0.00b | 615.00ms | +35.00ms ⚠️ |
| dist/clipboard.6a6e5fbd.js | 3.00kb | +0.00b | 729.00ms | +45.00ms ⚠️ |
| dist/ResourcedEmojiComponent.5046c0ff.js | 2.19kb | +0.00b | 718.00ms | +63.00ms ⚠️ |
| dist/card.2888338b.js | 2.15kb | +0.00b | 726.00ms | +41.00ms ⚠️ |
| dist/date.a69000aa.js | 1.97kb | +0.00b | 669.00ms | +46.00ms ⚠️ |
| dist/images.ece03ae5.js | 1.91kb | +0.00b | 670.00ms | +45.00ms ⚠️ |
| dist/feedback.facaac2e.js | 1.87kb | +0.00b | 710.00ms | +63.00ms ⚠️ |
| dist/16.a9ea9b41.js | 1.86kb | +0.00b | 661.00ms | +55.00ms ⚠️ |
| dist/browser.8883c452.js | 1.85kb | +0.00b | 729.00ms | +45.00ms ⚠️ |
| dist/workerHasher.31326e89.js | 1.77kb | +0.00b | 710.00ms | +47.00ms ⚠️ |
| dist/workerHasher.d607e0b8.js | 1.77kb | +0.00b | 729.00ms | +39.00ms ⚠️ |
| dist/list-number.d4568b49.js | 1.69kb | +0.00b | 673.00ms | +42.00ms ⚠️ |
| dist/status.6595d133.js | 1.69kb | +0.00b | 685.00ms | +38.00ms ⚠️ |
| dist/code.efde0f48.js | 1.61kb | +0.00b | 669.00ms | +47.00ms ⚠️ |
| dist/link.5be155be.js | 1.54kb | +0.00b | 673.00ms | +47.00ms ⚠️ |
| dist/heading6.5a6082c8.js | 1.53kb | +0.00b | 710.00ms | +63.00ms ⚠️ |
| dist/heading3.1c92891c.js | 1.52kb | +0.00b | 710.00ms | +64.00ms ⚠️ |
| dist/16.07ccad90.js | 1.51kb | +0.00b | 653.00ms | +47.00ms ⚠️ |
| dist/16.02056ae6.js | 1.47kb | +0.00b | 653.00ms | +35.00ms ⚠️ |
| dist/emoji.b5624cf5.js | 1.46kb | +0.00b | 670.00ms | +45.00ms ⚠️ |
| dist/16.00100306.js | 1.44kb | +0.00b | 669.00ms | +50.00ms ⚠️ |
| dist/16.bdd61d2d.js | 1.41kb | +0.00b | 662.00ms | +43.00ms ⚠️ |
| dist/heading5.88bc6249.js | 1.41kb | +0.00b | 710.00ms | +63.00ms ⚠️ |
| dist/expand.4a483eb0.js | 1.39kb | +0.00b | 709.00ms | +63.00ms ⚠️ |
| dist/16.fa5f411c.js | 1.36kb | +0.00b | 661.00ms | +43.00ms ⚠️ |
| dist/heading2.22de70fe.js | 1.34kb | +0.00b | 685.00ms | +39.00ms ⚠️ |
| dist/16.d10db5a3.js | 1.32kb | +0.00b | 661.00ms | +43.00ms ⚠️ |
| dist/16.c9db7e6b.js | 1.32kb | +0.00b | 669.00ms | +46.00ms ⚠️ |
| dist/mention.c44b48ed.js | 1.30kb | +0.00b | 673.00ms | +42.00ms ⚠️ |
| dist/heading4.a2524c70.js | 1.29kb | +0.00b | 710.00ms | +63.00ms ⚠️ |
| dist/Modal.44e980e0.js | 1.29kb | +0.00b | 627.00ms | +36.00ms ⚠️ |
| dist/layout.81f2196c.js | 1.28kb | +0.00b | 670.00ms | +45.00ms ⚠️ |
| dist/16.acff901d.js | 1.27kb | +0.00b | 653.00ms | +47.00ms ⚠️ |
| dist/component.233c7a65.js | 1.26kb | +0.00b | 667.00ms | +79.00ms ⚠️ |
| dist/16.828b958e.js | 1.26kb | +0.00b | 653.00ms | +47.00ms ⚠️ |
| dist/16.b9b1643b.js | 1.26kb | +0.00b | 653.00ms | +47.00ms ⚠️ |
| dist/divider.ed7ccd2e.js | 1.26kb | +0.00b | 670.00ms | +45.00ms ⚠️ |
| dist/quote.b6d8bffb.js | 1.25kb | +0.00b | 678.00ms | +45.00ms ⚠️ |
| dist/media-card-analytics-error-boundary.ff0177b3.js | 1.25kb | +0.00b | 716.00ms | +42.00ms ⚠️ |
| dist/action.01ec8983.js | 1.24kb | +0.00b | 669.00ms | +47.00ms ⚠️ |
| dist/16.458399e0.js | 1.23kb | +0.00b | 667.00ms | +87.00ms ⚠️ |
| dist/decision.c48f9f1e.js | 1.22kb | +0.00b | 670.00ms | +41.00ms ⚠️ |
| dist/panel-warning.b5f3360d.js | 1.22kb | +0.00b | 677.00ms | +47.00ms ⚠️ |
| dist/list.2dfe3f18.js | 1.19kb | +0.00b | 673.00ms | +42.00ms ⚠️ |
| dist/heading1.cc82b077.js | 1.19kb | +0.00b | 685.00ms | +39.00ms ⚠️ |
| dist/16.90358f40.js | 1.18kb | +0.00b | 662.00ms | +44.00ms ⚠️ |
| dist/panel-error.7454ebc0.js | 1.12kb | +0.00b | 678.00ms | +46.00ms ⚠️ |
| dist/panel.fc70749c.js | 1.11kb | +0.00b | 677.00ms | +47.00ms ⚠️ |
| dist/table.77df779b.js | 1.10kb | +0.00b | 685.00ms | +39.00ms ⚠️ |
| dist/panel-success.d5752da0.js | 1.06kb | +0.00b | 677.00ms | +48.00ms ⚠️ |
| dist/panel-note.6a71a1f2.js | 1.05kb | +0.00b | 677.00ms | +48.00ms ⚠️ |
| dist/media-picker-analytics-error-boundary.49b32f4d.js | 1023.00b | +0.00b | 725.00ms | +41.00ms ⚠️ |
| dist/media-card-analytics-error-boundary.9a500f44.js | 1019.00b | +0.00b | 716.00ms | +42.00ms ⚠️ |
| dist/media-card-analytics-error-boundary.b884d580.js | 1019.00b | +0.00b | 729.00ms | +39.00ms ⚠️ |
| dist/media-viewer-analytics-error-boundary.83da5db2.js | 1015.00b | +0.00b | 726.00ms | +41.00ms ⚠️ |
| dist/simpleHasher.79137cb4.js | 767.00b | +0.00b | 718.00ms | +63.00ms ⚠️ |
| dist/simpleHasher.2c5206fd.js | 767.00b | +0.00b | 733.00ms | +43.00ms ⚠️ |
| dist/index.html | 119.00b | +0.00b | 664.00ms | +146.00ms ⚠️ |

</p></details>

<details><summary>Three.js ✅</summary><p>

#### Timings

| Description | Time | Difference |
| --- | --- | --- |
| Cold | 21.55s | -197.00ms |
| Cached | 577.00ms | -61.00ms 🚀 |

#### Cold Bundles

*No bundle changes detected.*

#### Cached Bundles

*No bundle changes detected.*

</p></details>

**[Click here to view a detailed benchmark overview.](https://benchmark.parceljs.org/benchmark/291999605214675465)**

**Comment by mischnic** [comment]
Is this the same bug as https://github.com/parcel-bundler/parcel/issues/5181#issuecomment-699690550? The stack trace looks similar

**Comment by wbinnssmith** [comment]
> Is this the same bug as #5181 (comment)? The stack trace looks similar

Yeah, this prevents that exception from being thrown. However, the worker3 bundle is now successfully removed (it attempted to before), which isn't the desired behavior for workers. I pushed a change which properly keeps the worker bundle around.

**Review by joeyslater — APPROVED** [review]
Nice catch on the empty shared empty bundle

**Pull request merged**